pkg/api/v1/resource/helpers.go

```diff
@@ -83,20 +85,43 @@ func PodRequests(pod *v1.Pod, opts PodResourcesOptions) v1.ResourceList {
 		addResourceList(reqs, containerReqs)
 	}
 
+	var sidecarResources v1.ResourceList
+	if opts.SidecarContainersEnabled {
+		sidecarResources = make(v1.ResourceList)
+	}
+	initContainerReqs := v1.ResourceList{}
 	// init containers define the minimum of any resource
 	// Note: In-place resize is not allowed for InitContainers, so no need to check for ResizeStatus value
 	for _, container := range pod.Spec.InitContainers {
-		containerReqs := container.Resources.Requests
+		containerReqs := container.Resources.Requests.DeepCopy()
 		if len(opts.NonMissingContainerRequests) > 0 {
 			containerReqs = applyNonMissing(containerReqs, opts.NonMissingContainerRequests)
 		}
 
+		if opts.SidecarContainersEnabled {
+			if container.RestartPolicy != nil && *container.RestartPolicy == v1.ContainerRestartPolicyAlways {
```

Comment:
I believe that one of the main reasons for gating the feature is to minimize its impact, even if it seems obvious.

Reply:
@jpbetz 
Re-thinking about this in terms of https://github.com/kubernetes/kubernetes/pull/116429#discussion_r1189957266,

This is more tricky as it affects the kube-scheduler as well and may have real effects not just printing the information.

What do you think?

Reply:
> I don't see any harm to reporting the resources limits/requests accurately on sidecar pods when enforcing all this.

@tzneal 

What do you think? (as you know more context about resource calculation use cases)

Reply:
I *think* it may be ok to remove the feature flag check and always calculate the resources with sidecars, assuming that the scheduler and kubelet reject those pods when the flag is disabled. 

The scheduler rejecting the pod guards against a sort of weird scenario where a 1.28 scheduler with the flag disabled could schedule the pod against a 1.27 kubelet with no knowledge of the flag. In that case the kublet and kube-scheduler would calculate different opinions of the amount of resources being consumed.

Reply:
Sounds good. I'll defer to you both on which approach seems the lowest risk.

Reply:
Please open a KEP PR with the proposed solution.

I agree that for scheduled pods, the calculations should respect the sidecar containers.

For unscheduled pods, I don't recall an existing case where we block pod scheduling due to additional fields. I'm not keen to the idea of having this validation in scheduler that will eventually be removed. But if the idea is to stop any kind of bleeding, it makes sense for the NodeResourcesFit plugin to report Unschedulable.

Reply:
added a section here: https://github.com/kubernetes/enhancements/pull/3968/files

> #### Resource calculation and version skew
> 
> In case of a version skew between scheduler and kubeler, or in cases when
> scheduler and kubelet has a different value set for the `Sidecar` feature gate,
> calculation of resources required for a Pod will differ between the scheduler
> and a kubelet when the sidecar container created.
> 
> In case when scheduler "knows" about the sidecar and kubelet doesn't, there
> unlikely be any issues. Scheduler will calculate resources usage for a Pod that
> will be equal or more than kubelet will require to run the Pod. So there will be
> no overbooking.
> 
> If scheduler has the `Sidecar` feature gate disabled, the Pod that has a Sidecar
> container will not be admitted as validation of the new field will fail.
> 
> We will recommend in documentation to not disable the feature gate on scheduler,
> while there are any Pods with Sidecar container is running.

Reply:
https://github.com/kubernetes/kubernetes/pull/116429/commits/4a0522df49e946d2d2c9bcde3aedfd71cb4f238e

added validation to NodeResourceFit plugin according to the latest KEP.